### PR TITLE
drivers: sensor: Add HS400x driver

### PR DESCRIPTION
**Pull request by TotallyOriginalUsername**
Adds Renesas HS400x temperature and humidity sensors driver.
Signed-off-by: Mariëlle Korthout <marielle.korthout.jc@renesas.com>

### CODE REVIEW

**Review by MaureenHelm — COMMENTED** [review]
Thanks for the contribution!

This looks very similar to the `hs300x`. Could that driver be enhanced instead of introducing a separate driver?

Please have a look at the failed compliance check and address the error.

**Comment by TotallyOriginalUsername** [comment]
The driver indeed looks similar and I have been able to use a lot of the hs300x driver, but the hs400x requires an extra measurement command byte and sends back an extra CRC byte. 
I think it is better to introduce a separate hs400x driver for those differences in communication and so the driver can later be enhanced to support other modes of the hs400x which the hs300x does not have.

File: drivers/sensor/renesas/hs400x/hs400x.c
Comment (kartben) [inline drivers/sensor/renesas/hs400x/hs400x.c:48]:
```suggestion
	 * The sensor sends a checkum after each measurement. See datasheet "CRC Checksum
```

File: dts/bindings/sensor/renesas,hs400x.yaml
Comment (kartben) [inline dts/bindings/sensor/renesas,hs400x.yaml:2]: Copyright who?

File: tests/drivers/build_all/sensor/i2c.dtsi
Comment (kartben) [inline tests/drivers/build_all/sensor/i2c.dtsi:586]: You need to add the sensor at the end of the list, at the next available address (here 54 is already used by another sensor)

File: drivers/sensor/renesas/hs400x/Kconfig
Comment (kartben) [inline drivers/sensor/renesas/hs400x/Kconfig:9]:
```suggestion
 depends on DT_HAS_RENESAS_HS400X_ENABLED
 select CRC
```

There are many sensors that make CRC check optional and something that can be enabled/disabled via ab additional Kconfigs option. I'll let sensor maintainers comment on whether that's preferable to forcing users to pull CRC code if they don't want or can't afford it (it feels to me like it would indeed be better, and is likely a reasonably trivial change)

File: drivers/sensor/renesas/hs400x/hs400x.c
Comment (kartben) [inline drivers/sensor/renesas/hs400x/hs400x.c:53]:
```suggestion
		LOG_ERR("CRC check failed: computed=%u,expected=%u", crc, rx_buf[4]);
```

**Review by kartben — CHANGES_REQUESTED** [review]

File: drivers/sensor/renesas/hs400x/Kconfig
Comment (TotallyOriginalUsername) [reply drivers/sensor/renesas/hs400x/Kconfig:9]: It also seems more preferable to me to allow the user to decide, thank you for the suggestion. I've added a Kconfig option for this.

File: tests/drivers/build_all/sensor/i2c.dtsi
Comment (kartben) [inline tests/drivers/build_all/sensor/i2c.dtsi:1091]: please rebase on main: this should be `@9f`at the time of writing (and not sure why there is a bogus apds_9306 entry, looks like it was fixed at some point though https://github.com/zephyrproject-rtos/zephyr/commit/5be36eef47c006a4238400df79d6b667cc809b58#diff-e5e68be0d8d187414fd7e5acd16a07df950cace9e49bac77f4da6846dfe8a505R1106)

File: drivers/sensor/renesas/hs400x/hs400x.c
Comment (pdgendt) [inline drivers/sensor/renesas/hs400x/hs400x.c:154]: Use the recently introduced `DEVICE_API` wrapper and add a trailing comma to align this block to the left.

```suggestion
static DEVICE_API(sensor, hs400x_driver_api) = {
	.sample_fetch = hs400x_sample_fetch,
	.channel_get = hs400x_channel_get,
};
```

File: tests/drivers/build_all/sensor/i2c.dtsi
Comment (TotallyOriginalUsername) [reply tests/drivers/build_all/sensor/i2c.dtsi:1091]: Thank you, I've rebased on main (not sure why either, I worked from the 3.7 release)

File: drivers/sensor/renesas/hs400x/hs400x.c
Comment (TotallyOriginalUsername) [reply drivers/sensor/renesas/hs400x/hs400x.c]: Thank you, it's now using the DEVICE_API wrapper

File: drivers/sensor/renesas/hs400x/hs400x.c
Comment (TotallyOriginalUsername) [reply drivers/sensor/renesas/hs400x/hs400x.c]: Alright, I've removed the guarding there

**Review by kartben — DISMISSED** [review]

**Comment by pdgendt** [comment]
Please check CI issues

File: tests/drivers/build_all/sensor/i2c.dtsi
Comment (kartben) [inline tests/drivers/build_all/sensor/i2c.dtsi:1155]:
```suggestion
test_i2c_hs400x: hs400x@a0 {
	compatible = "renesas,hs400x";
	reg = <0xa0>;
};
```

Comment (TotallyOriginalUsername) [reply tests/drivers/build_all/sensor/i2c.dtsi]: Adjusted, thanks! I totally overlooked the capitalization and missing > yesterday

**Review by kartben — DISMISSED** [review]

File: drivers/sensor/renesas/hs400x/Kconfig
Comment (pdgendt) [inline drivers/sensor/renesas/hs400x/Kconfig:17]: Leave out `default n`, it's the default.

File: drivers/sensor/renesas/hs400x/hs400x.c
Comment (pdgendt) [inline drivers/sensor/renesas/hs400x/hs400x.c:51]: Please do not indent `#if` directive.

File: drivers/sensor/renesas/hs400x/hs400x.c
Comment (pdgendt) [inline drivers/sensor/renesas/hs400x/hs400x.c:56]: Returning `crc` seems wrong.

```suggestion
		return -EIO;
```

**Review by pdgendt — DISMISSED** [review]

**Review by pdgendt — APPROVED** [review]
> This looks very similar to the `hs300x`. Could that driver be enhanced instead of introducing a separate driver?

I agree here, but it's better to have something rather than nothing 🤷

File: tests/drivers/build_all/sensor/i2c.dtsi
Comment (TotallyOriginalUsername) [reply tests/drivers/build_all/sensor/i2c.dtsi:1191]: Thank you for your reviews, I've moved the addition to the tests in a separate commit

**Review by kartben — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**
